Implemented Context Menus

drobbins329 opened the pull request:
Added ApplicationCommand types:
Slash : 1
User: 2
Message: 3

And the appropriate CRUD methods.

Thanks in advance for your contribution to Discord.Net!

quinchs committed:
Update README.md

quinchs committed:
Update README.md

Rutherther committed:
Fix SocketSlashCommandDataOption to use long for Number instead of int (#89)

drobbins329 committed:
Application webhooks (#86)

* Added webhook components for hooks having an application ID.

quinchs committed:
resolved #88

quinchs committed:
resolved #85

quinchs committed:
Update device for gateway

d4n3436 committed:
Fix MessageProperties.Embed being ignored in some methods that modifies a message (#92)

MrCakeSlayer committed:
Update label/description lengths for selects (ref: https://github.com/discord/discord-api-docs/pull/3598/files) (#91)

https://github.com/discord/discord-api-docs/pull/3598/files

d4n3436 committed:
Fix tests (#90)

Rutherther committed:
Fix gateway serialization to include nulls (#96)

MrCakeSlayer committed:
Add missing guild permissions (#93)

* Update GuildPermissions.cs

* Update GuildPermissionsTests.cs

INikonI committed:
Add banner and accent color to user and some fixes/improvements (#81)

* Add banner and accent color to user and some fixes

* Fix

* Fix!

* increase size of user banners to 256

* Some changes and mini refactor of color class

* add constant maxDecimalValue to color and checks with exceptions

* add `NotSupportedException` for `BannerId` and `AccentColor` in `SocketWebhookUser`

* Update ComponentBuilder.cs

- `MaxLabelLength` from `ComponentBuilder` moved to `ButtonBuilder`
- Added `MaxLabelLength` for `SelectMenuOptionBuilder`
- Changed `MaxDescriptionLength` to 100

Cenngo committed:
Interface Method Declarations for Interaction Methods (#99)

* added interface method declarations

* inline docs

quinchs committed:
Fix serialization error

quinchs committed:
Merge remote-tracking branch 'origin/release/2.x' into release/3.x

quinchs committed:
meta: bump versions

quinchs committed:
Fix debug pragma

quinchs committed:
meta: bump version

quinchs committed:
Remove rich presence button

Cenngo committed:
Assign CurrentUserId in Sharded Client (#100)

* added interface method declarations

* inline docs

* current user id assignment in sharded client

quinchs committed:
Merge remote-tracking branch 'origin/feature/interations' into release/3.x

d4n3436 committed:
Allow EmbedBuilder.ImageUrl to use attachment scheme syntax (#104)

drobbins329 commented:
Did not implement any of the new interaction created objects. This build currently only supports receiving them as slash commands.
edit: implemented in last commit.

drobbins329 commented:
Added commit. 
Brings SocketApplicationUserCommand, containing *who* was clicked and SocketApplicationMessageCommand, containing *what message* was clicked.

Rutherther commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -235,6 +235,7 @@ private async Task LogoutInternalAsync()
             options.BucketId = bucketId;
 
             string json = payload != null ? SerializeJson(payload) : null;
+            Console.WriteLine($"Sending JSON....\n{json}");
```
I feel like this is a forgotten line for debugging purposes? But it could be useful if it was wrapped in preprocessor #if check with something like DEBUG_REST

drobbins329 replied:
thats exactly what it was. removed.

Rutherther commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1163,6 +1216,87 @@ public async Task<ApplicationCommand[]> BulkOverwriteGuildApplicationCommands(ul
+
+            var bucket = new BucketIds(guildId: guildId);
+
+            try
+            {
+                return await SendJsonAsync<ApplicationCommand>("PATCH", () => $"applications/{this.CurrentUserId}/guilds/{guildId}/commands/{commandId}", command, bucket, options: options).ConfigureAwait(false);
+            }
+            catch (HttpException x)
+            {
+                if (x.HttpCode == HttpStatusCode.BadRequest)
+                {
+                    var json = (x.Request as JsonRestRequest).Json;
+                    throw new ApplicationCommandException(json, x);
+                }
+
+                // Re-throw the http exception
+                throw;
```
Is there a reason why you didn't use `TrySendApplicationCommand`? It seems like it does the same thing as this try catch

Same goes for `ModifyGuildApplicationMessageCommandAsync`, `ModifyGuildApplicationCommandAsync` (I know this one wasn't written by you)

Rutherther commented on src/Discord.Net.Rest/Entities/Interactions/RestGuildMessageCommand.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    ///     Represents a Rest-based guild command.
+    /// </summary>
+    public class RestGuildMessageCommand : RestApplicationCommand
```
As now there are more guild commands, wouldn't it be better to have interface or base class for guild commands that would have `GetGuild` method? That way underlying guild of command could be easier to get from yet unknown type of command instead of checking for Message, User, Slash commands

quinchs replied:
I agree with this

drobbins329 replied:
So, duplicate what i did with the SocketCommandBase, But as a RestGuildCommandBase, just to implement the GetGuild method?

Rutherther commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/User Commands/SocketApplicationUserCommand.cs:
```diff
@@ -0,0 +1,165 @@
+            if (!IsValidToken)
+                throw new InvalidOperationException("Interaction token is no longer valid");
+
+            if (embeds == null && embed != null)
+                embeds = new[] { embed };
+            Preconditions.AtMost(allowedMentions?.RoleIds?.Count ?? 0, 100, nameof(allowedMentions.RoleIds), "A max of 100 role Ids are allowed.");
+            Preconditions.AtMost(allowedMentions?.UserIds?.Count ?? 0, 100, nameof(allowedMentions.UserIds), "A max of 100 user Ids are allowed.");
+            Preconditions.AtMost(embeds?.Length ?? 0, 10, nameof(embeds), "A max of 10 embeds are allowed.");
+
+            var args = new API.Rest.CreateWebhookMessageParams
+            {
+                Content = text,
+                AllowedMentions = allowedMentions?.ToModel() ?? Optional<API.AllowedMentions>.Unspecified,
+                IsTTS = isTTS,
+                Embeds = embeds?.Select(x => x.ToModel()).ToArray() ?? Optional<API.Embed[]>.Unspecified,
+                Components = component?.Components.Select(x => new API.ActionRowComponent(x)).ToArray() ?? Optional<API.ActionRowComponent[]>.Unspecified
+            };
+
+            if (ephemeral)
+                args.Flags = 64;
+
+            return await InteractionHelper.SendFollowupAsync(Discord.Rest, args, Token, Channel, options);
+        }
+
+        /// <summary>
+        ///     Acknowledges this interaction with the <see cref="InteractionResponseType.DeferredChannelMessageWithSource"/>.
+        /// </summary>
+        /// <returns>
+        ///     A task that represents the asynchronous operation of acknowledging the interaction.
+        /// </returns>
+        public override Task DeferAsync(RequestOptions options = null)
+        {
+            var response = new API.InteractionResponse
+            {
+                Type = InteractionResponseType.DeferredChannelMessageWithSource,
+            };
+
+            return Discord.Rest.ApiClient.CreateInteractionResponse(response, this.Id, this.Token, options);
+        }
+    }
```
Is there any difference between these and the ones in base class `SocketSlashCommand`? I don't see any, so if there aren't what is the reason for overwriting them and not using the ones inherited from the base class? Same goes for `SocketApplicationMessageCommand`

Rutherther commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/Message Commands/SocketApplicationMessageCommand.cs:
```diff
@@ -0,0 +1,165 @@
+        {
+            if (!IsValidToken)
+                throw new InvalidOperationException("Interaction token is no longer valid");
+
+            if (embeds == null && embed != null)
+                embeds = new[] { embed };
+            Preconditions.AtMost(allowedMentions?.RoleIds?.Count ?? 0, 100, nameof(allowedMentions.RoleIds), "A max of 100 role Ids are allowed.");
+            Preconditions.AtMost(allowedMentions?.UserIds?.Count ?? 0, 100, nameof(allowedMentions.UserIds), "A max of 100 user Ids are allowed.");
+            Preconditions.AtMost(embeds?.Length ?? 0, 10, nameof(embeds), "A max of 10 embeds are allowed.");
+
+            var args = new API.Rest.CreateWebhookMessageParams
+            {
+                Content = text,
+                AllowedMentions = allowedMentions?.ToModel() ?? Optional<API.AllowedMentions>.Unspecified,
+                IsTTS = isTTS,
+                Embeds = embeds?.Select(x => x.ToModel()).ToArray() ?? Optional<API.Embed[]>.Unspecified,
+                Components = component?.Components.Select(x => new API.ActionRowComponent(x)).ToArray() ?? Optional<API.ActionRowComponent[]>.Unspecified
+            };
+
+            if (ephemeral)
+                args.Flags = 64;
+
+            return await InteractionHelper.SendFollowupAsync(Discord.Rest, args, Token, Channel, options);
+        }
+
+        /// <summary>
+        ///     Acknowledges this interaction with the <see cref="InteractionResponseType.DeferredChannelMessageWithSource"/>.
+        /// </summary>
+        /// <returns>
+        ///     A task that represents the asynchronous operation of acknowledging the interaction.
+        /// </returns>
+        public override Task DeferAsync(RequestOptions options = null)
+        {
+            var response = new API.InteractionResponse
+            {
+                Type = InteractionResponseType.DeferredChannelMessageWithSource,
+            };
+
+            return Discord.Rest.ApiClient.CreateInteractionResponse(response, this.Id, this.Token, options);
+        }
```
See comment in `SocketApplicationUserCommand` file

Rutherther commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1163,6 +1216,87 @@ public async Task<ApplicationCommand[]> BulkOverwriteGuildApplicationCommands(ul
+
+            var bucket = new BucketIds(guildId: guildId);
+
+            try
+            {
+                return await SendJsonAsync<ApplicationCommand>("PATCH", () => $"applications/{this.CurrentUserId}/guilds/{guildId}/commands/{commandId}", command, bucket, options: options).ConfigureAwait(false);
+            }
+            catch (HttpException x)
+            {
+                if (x.HttpCode == HttpStatusCode.BadRequest)
+                {
+                    var json = (x.Request as JsonRestRequest).Json;
+                    throw new ApplicationCommandException(json, x);
+                }
+
+                // Re-throw the http exception
+                throw;
+            }
```
See comment for `ModifyGuildApplicationUserCommandAsync`

Rutherther reviewed (commented):
I have taken a look at the code and left some ideas in it.

It also doesn't feel right to me to just inherit like this `SocketApplicationMessageCommand : SocketSlashCommand`. AFAIK base type `SocketSlashCommand` is still for commands of type slash. So wouldn't it make more sense to have a `SocketCommandBase` class and then class for Message, User and Slash command types? and not inherit from Slash command in Message and User commands?

drobbins329 commented:
> I have taken a look at the code and left some ideas in it.
> 
> It also doesn't feel right to me to just inherit like this `SocketApplicationMessageCommand : SocketSlashCommand`. AFAIK base type `SocketSlashCommand` is still for commands of type slash. So wouldn't it make more sense to have a `SocketCommandBase` class and then class for Message, User and Slash command types? and not inherit from Slash command in Message and User commands?

implemented. moved all the response/followup/and defer methods to the base.

Rutherther commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBase.cs:
```diff
@@ -0,0 +1,164 @@
+
+        new internal static SocketInteraction Create(DiscordSocketClient client, Model model, ISocketMessageChannel channel)
+        {
+            var entity = new SocketSlashCommand(client, model, channel);
```
Shouldn't this be implemented in each inherited class separately? this way SocketSlashCommand will be created for slash, message and user commands, won't it?

drobbins329 replied:
oops. fixed. thanks.

Rutherther commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/User Commands/SocketApplicationUserCommandData.cs:
```diff
@@ -6,7 +6,7 @@
-    ///     Represents the data tied with the <see cref="SocketSlashCommand"/> interaction.
+    ///     Represents the data tied with the <see cref="SocketApplicationUserCommand"/> interaction.
     /// </summary>
     public class SocketApplicationUserCommandData : SocketEntity<ulong>, IApplicationCommandInteractionData
```
Shouldn't this inherit from BaseData? Or is the purpose of that class different?

drobbins329 replied:
slightly different. one is an interface, the other is a class to support the SocketCommandBase methods. Namely, the Update mthod.

Rutherther commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/Message Commands/SocketApplicationMessageCommandData.cs:
```diff
@@ -6,7 +6,7 @@
-    ///     Represents the data tied with the <see cref="SocketSlashCommand"/> interaction.
+    ///     Represents the data tied with the <see cref="SocketApplicationMessageCommand"/> interaction.
     /// </summary>
     public class SocketApplicationMessageCommandData : SocketEntity<ulong>, IApplicationCommandInteractionData
```
Shouldn't this inherit from BaseData? Or is the purpose of that class different?

drobbins329 replied:
slightly different. one is an interface, the other is a class to support the SocketCommandBase methods. Namely, the Update mthod.

MrCakeSlayer commented:
@drobbins329 Could you also include updated information in the README/Docs with usage information for the new context menus in labs?

drobbins329 commented:
Should the read me be an addition to the slash commands one? As they're related?

Move the slash commands guide into an Application Commands parent folder. The add User commands and Message Commands along side it?

drobbins329 commented:
Moved the slash commands down one. Parent directory is now "Application Commands". I've gotten the creation of the context menu commands into their own guide page now. Never built a github guide before, so struggling a little on where to go next.

Rutherther committed:
Make Webhook ApplicationId nullable instead of optional + fix IDiscordInteraction DeferAsync method (#110)

* Make Webhook ApplicationId nullable instead of optional

* Fix IDiscordInteraction DeferAsync to account for ephemeral defer

quinchs commented on docs/guides/application-commands/context-menu-commands/receiving-context-menu-command-events.md:
```diff
@@ -0,0 +1,43 @@
+	if(slashCommand != null)
+		Console.Writeline("Slash command received!")
+		
+	var userCommand = arg as SocketApplicationUserCommand;
```
Maybe do a switch statement for the types as it looks a bit nicer and is easier to read

drobbins329 replied:
Unsure how to proceed with this.

Rutherther replied:
```
switch (arg) {
  case SocketApplicationUserCommand userCommand:
    // ...
    break;
  // ...
}
```

drobbins329 replied:
i.... did not know this was a thing, honestly. So, it will only work if the cast results in non-null?

Rutherther replied:
It will check the type and if it matches cast to it, then it will be accessible as `userCommand` local property.

Also `if (smth is SomeType someType)` is useful instead of as and then checking for null, but if there are more types, switch is better

quinchs commented on src/Discord.Net.Core/Entities/Interactions/ApplicationCommandTypes.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Discord
+{
+    public enum ApplicationCommandType : byte
```
Add summary tags for public members

drobbins329 replied:
done

quinchs commented on src/Discord.Net.Core/Entities/Interactions/MessageCommandCreationProperties.cs:
```diff
@@ -0,0 +1,30 @@
+    /// <summary>
+    ///     A class used to create Message commands.
+    /// </summary>
+    public class MessageCommandCreationProperties
```
Remove unnecessary properties like the type as the type is inferred from the name of the class. Also the description is not used?

drobbins329 replied:
Description property removed. Type is used in command creation, and command interaction.

quinchs replied:
Yes but you have specified 2 different classes for application commands, user and message. Shouldnt the type be inferred inside of the api method and set their?

drobbins329 replied:
you're saying set the type in DiscordRestApiClient for the create/modify for user/message/slash commands and just use the class names on receive? (WebSocket/.../SocketInteraction.cs)

So don't bother with the type in the SocketUser/Message/SlashCommand? And leave it solely in the ApplicationCommandInteractionData.cs?

quinchs commented on src/Discord.Net.Core/Entities/Interactions/UserCommandCreationProperties.cs:
```diff
@@ -0,0 +1,30 @@
+        /// </summary>
+        public string Description { get; set; }
+
+
```
Remove double newline

drobbins329 replied:
File removed.

quinchs commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1095,6 +1108,46 @@ public async Task<ApplicationCommand[]> BulkOverwriteGlobalApplicationCommands(C
+
+            options = RequestOptions.CreateOrClone(options);
+
+            
```
remove newlines

drobbins329 replied:
done

quinchs commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1095,6 +1108,46 @@ public async Task<ApplicationCommand[]> BulkOverwriteGlobalApplicationCommands(C
+
+            options = RequestOptions.CreateOrClone(options);
+
+
```
remove newlines

quinchs commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1095,6 +1108,46 @@ public async Task<ApplicationCommand[]> BulkOverwriteGlobalApplicationCommands(C
+            return await TrySendApplicationCommand(SendJsonAsync<ApplicationCommand[]>("PUT", () => $"applications/{this.CurrentUserId}/commands", commands, new BucketIds(), options: options)).ConfigureAwait(false);
+        }
+
+        public async Task<ApplicationCommand[]> BulkOverwriteGlobalApplicationMessageCommands(CreateApplicationCommandParams[] commands, RequestOptions options = null)
```
Duplicate methods? Should just be `BulkOverwriteGlobalApplicationCommands`

quinchs commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1163,6 +1216,87 @@ public async Task<ApplicationCommand[]> BulkOverwriteGuildApplicationCommands(ul
+            var bucket = new BucketIds(guildId: guildId);
+
+            return await TrySendApplicationCommand(SendJsonAsync<ApplicationCommand>("POST", () => $"applications/{this.CurrentUserId}/guilds/{guildId}/commands", command, bucket, options: options)).ConfigureAwait(false);
+
```
remove newline

quinchs commented on src/Discord.Net.Rest/DiscordRestApiClient.cs:
```diff
@@ -1163,6 +1216,87 @@ public async Task<ApplicationCommand[]> BulkOverwriteGuildApplicationCommands(ul
+            return await TrySendApplicationCommand(SendJsonAsync<ApplicationCommand[]>("PUT", () => $"applications/{this.CurrentUserId}/guilds/{guildId}/commands", commands, bucket, options: options)).ConfigureAwait(false);
+        }
+
+        public async Task<ApplicationCommand> CreateGuildApplicationMessageCommandAsync(CreateApplicationCommandParams command, ulong guildId, RequestOptions options = null)
```
Duplicate methods: rename to `CreateGuildApplicationCommandAsync`

quinchs commented on src/Discord.Net.Rest/Entities/Interactions/InteractionHelper.cs:
```diff
@@ -1,6 +1,7 @@
 using Discord.API;
 using Discord.API.Rest;
 using Discord.Net;
+//using Discord.Rest.Entities.Interactions;
```
Remove

quinchs commented on src/Discord.Net.Rest/Entities/Interactions/InteractionHelper.cs:
```diff
@@ -63,6 +64,7 @@ public static Task DeleteAllGlobalCommandsAsync(BaseDiscordClient client, Reques
             {
                 Name = arg.Name,
                 Description = arg.Description,
+                Type= arg.Type,
```
Add space before `=`

quinchs commented on src/Discord.Net.Rest/Entities/Interactions/InteractionHelper.cs:
```diff
@@ -195,6 +200,174 @@ public static async Task DeleteGlobalCommand(BaseDiscordClient client, RestGloba
+            return await CreateGlobalUserCommand(client, args, options).ConfigureAwait(false);
+        }
+
+        public static async Task<RestGlobalUserCommand> CreateGlobalUserCommand(BaseDiscordClient client, UserCommandCreationProperties arg, RequestOptions options = null)
```
Please simplify by creating a `ApplicationCommandProperties` that can cover both user and message commands, then remove the duplicate methods

quinchs commented on src/Discord.Net.Rest/Entities/Interactions/RestApplicationCommandType.cs:
```diff
@@ -15,10 +15,14 @@ public enum RestApplicationCommandType
         ///     Specifies that this command is a Global command.
         /// </summary>
         GlobalCommand,
+        GlobalUserCommand,
```
Add `summary` tags for public members

quinchs commented on src/Discord.Net.Rest/Entities/Interactions/RestApplicationCommandType.cs:
```diff
@@ -15,10 +15,14 @@ public enum RestApplicationCommandType
         ///     Specifies that this command is a Guild specific command.
         /// </summary>
-        GuildCommand
+        GuildCommand,
```
Add `summary` tags for public members

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/Message Commands/SocketApplicationMessageCommandData.cs:
```diff
@@ -0,0 +1,140 @@
+    {
+        /// <inheritdoc/>
+        public string Name { get; private set; }
+        public SocketMessage Message {  get; private set; }
```
Add `summary` tags for public members

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/Message Commands/SocketApplicationMessageCommandData.cs:
```diff
@@ -0,0 +1,140 @@
+        public string Name { get; private set; }
+        public SocketMessage Message {  get; private set; }
+
+        internal Dictionary<ulong, SocketGuildUser> guildMembers { get; private set; }
```
For resolving types there should be a base `SocketApplicationCommandData` class that handles resolvable types

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/User Commands/SocketApplicationUserCommandData.cs:
```diff
@@ -0,0 +1,113 @@
+        /// <inheritdoc/>
+        public string Name { get; private set; }
+
+        public SocketUser Member { get; private set; }
```
Add `summary` tag for public members

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/Context Menu Commands/User Commands/SocketApplicationUserCommandData.cs:
```diff
@@ -0,0 +1,113 @@
+
+        public SocketUser Member { get; private set; }
+
+        internal Dictionary<ulong, SocketGuildUser> guildMembers { get; private set; }
```
See other comment about resolving types

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseData.cs:
```diff
@@ -0,0 +1,145 @@
+
+namespace Discord.WebSocket
+{
+    public class SocketCommandBaseData : SocketEntity<ulong>, IApplicationCommandInteractionData
```
Add summary

drobbins329 replied:
done

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseData.cs:
```diff
@@ -0,0 +1,145 @@
+{
+    public class SocketCommandBaseData : SocketEntity<ulong>, IApplicationCommandInteractionData
+    {
+        public string Name { get; private set; }
```
Add summary

drobbins329 replied:
done

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseData.cs:
```diff
@@ -0,0 +1,145 @@
+    {
+        public string Name { get; private set; }
+
+        public IReadOnlyCollection<SocketCommandBaseDataOption> Options { get; private set; }
```
Add summary

drobbins329 replied:
done

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseDataOption.cs:
```diff
@@ -0,0 +1,130 @@
+
+namespace Discord.WebSocket
+{
+    public class SocketCommandBaseDataOption : IApplicationCommandInteractionDataOption
```
Add summary

quinchs commented on src/Discord.Net.WebSocket/Entities/Interaction/SocketBaseCommand/SocketCommandBaseDataOption.cs:
```diff
@@ -0,0 +1,130 @@
+{
+    public class SocketCommandBaseDataOption : IApplicationCommandInteractionDataOption
+    {
+        public string Name { get; private set; }
```
Add summary

quinchs reviewed (changes requested):
This looks good so far, but there is some stuff that needs changes.

- [ ] `UserCommandCreationProperties` and `MessageCommandCreationProperties` have the same properties yet different classes?
- [ ] Change `SocketApplicationMessageCommand` and `SocketApplicationUserCommand` to SocketMessageCommand and SocketUserCommand.repectively?
- [ ] Remove unneeded properties in `MessageCommandCreationProperties` and `UserCommandCreationProperties`.
- [ ] Some doccumentation is missing on some public properties.

quinchs commented:
Also please resolve conflicts

quinchs committed:
Fix application command and thread starter messages being created as SocketSystemMessage

drobbins329 committed:
Added description of ApplicationCommandType Enums

drobbins329 committed:
Requested Fixes

renamed SocketApplicationUserCommand to SocketUserCommand
renamed SocketApplicationMessageCommand to SocketMessageCommand
using ContextMenuCreationProperties for both User and Message commands

drobbins329 committed:
Added Summary to public members

removed whitespace from DiscordRestApiClient.cs

drobbins329 commented:
> Also please resolve conflicts

The SocketSlashCommand conflicts are due to moving the common respond/followup/etc methods to the SocketCommandBase.

drobbins329 commented:
> Also please resolve conflicts

Actually, both conflicts are. How would I resolve the conflicts, and leave the SocketCommandBase?

Rutherther commented:
I don't know what tools you are using for git, but best solution would be to rebase on release/3.x (meaning being checkouted in your branch, calling `git rebase release/3.x`), you will have to resolve the conflicts manually. By executing the command from command line you will have to go to the files, resolve it and then execute command to continue. If you are using tool like Sublime Merge it will guide you with the resolving.

After you are done resolving you will have to force push (because rebase changes git history). If you will be doing this, DOUBLE check that you didn't remove something important, because you won't be able to take it back if you don't make backups (like making separate branch or copying the files to another folder)

Easier solution is to merge release/3.x into your branch (calling `git merge release/3.x` from your branch), that way you don't have to force push (but it is not that clean as rebase, because new merge commit will be created). You will still have to resolve the conflicts.

Don't forget to first pull newest release/3.x commit into your local repository so you are rebasing on / merging the correct commit.

drobbins329 committed:
Fixing guide to use switch statement

drobbins329 committed:
Merge branch 'release/3.x' into feature/context_commands

drobbins329 committed:
implemented TrySendApplicationCommandAsync

drobbins329 committed:
implemented ephemeral in SocketCommandBase Defer, and RespondAsync.

assigning int 64 was error. changed to "MessageFlags.Ephemeral", built and tested to work.

quinchs commented:
Imma create a new branch for menus and merge this tmr, then im thinking about finishing the style changes and some code cleanup

drobbins329 committed:
removed ApplicationCommandType from SocketUser and SocketMessageCommandData

quinchs commented:
@drobbins329 Have a look at the [feature/context_commands](https://github.com/Discord-Net-Labs/Discord.Net-Labs/tree/feature/context_commands) branch and lmk what you think. Also have you joined the [dnet labs](https://discord.com/invite/dvSfUTet3K) discord server?
